src/extension_traits/vec.rs

```diff
@@ -0,0 +1,346 @@
+      -> Out<'_, [T]>
+    {
+        let capacity = self.capacity();
+        unsafe {
+            // # Safety
+            //
+            //   - `Vec<T>` safety invariant / layout guarantees state that
+            //     it owns a `Box<[MaybeUninit<T>]` of length `self.capacity()`
+            //     and starting at `self.as_mut_ptr()`.
+            slice::from_raw_parts_mut(
+                self.as_mut_ptr().cast::<MaybeUninit<T>>(),
+                capacity,
+            ).as_out()
+        }
+    }
+
+    /// Extracts an owned handle to the backing buffer.
+    #[inline]
+    fn into_backing_buffer (self: Vec<T>)
+      -> Box<[MaybeUninit<T>]>
+    where
+        T : Copy, // Opinionated stance against accidental memory leaks
+    {
+        self.into_backing_buffer_manually_drop()
+    }
+
+    /// Same as [`.into_backing_buffer()`][
+    /// `VecCapacity::into_backing_buffer`] but without the `Copy` bound.
+    ///
+    /// This means that extra care should be taken if
+    /// `mem::needs_drop::<Self::Item>()`. Indeed, the returned boxed slice
+    /// will not run the destructor of its initialized elements (since it no
+    /// longer knows which are).
+    ///
+    /// ⚠️ **Misusage of this function can thus lead to memory leaks** ⚠️
+    #[inline]
+    fn into_backing_buffer_manually_drop (self: Vec<T>)
+      -> Box<[MaybeUninit<T>]>
+    {
+        // We need to go through an intermediate max-capacity `Vec` because
+        // there is no official guarantee that a `Vec`'s backing allocation
+        // matches that of a boxed slice.
+        let this: Vec<MaybeUninit<T>> = unsafe {
+            // Safety:
+            //
+            //   - same layout;
+            //
+            //   - `MaybeUninit<MaybeUninit<_>>` is sound to `.assume_init()`;
+            //
+            //   - init -> uninit conversion is sound by "covariance of ownership".
+            let mut this = mem::ManuallyDrop::new(self);
+            let capacity = this.capacity();
+            Vec::from_raw_parts(
+                this.as_mut_ptr().cast::<MaybeUninit<T>>(),
+                capacity,
+                capacity,
+            )
+        };
+        this.into_boxed_slice() // hopefully a no-op
+    }
+}
+
+/// Extension trait for [`Vec`], that grows the vec by a _bounded_ amount of
+/// bytes, obtained when reading from `R`.
+///
+/// This guarantees that the allocated memory starts uninitialized (before
+/// being initialized when read), for maximum performance.
+///
+/// # Example
+///
+/// ```rust
+/// use ::uninit::read::VecExtendFromReader;
+///
+/// let mut reader = &b"World!"[..];
+/// let mut vec = b"Greetings, ".to_vec();
+/// vec.extend_from_reader(6, &mut reader).unwrap();
+/// assert_eq!(
+///     vec,
+///     b"Greetings, World!",
+/// );
+/// ```
+pub
+trait VecExtendFromReader<R : ReadIntoUninit> {
+    fn extend_from_reader (
+        self: &'_ mut Self,
+        count: usize,
+        reader: R,
+    ) -> io::Result<()>
+    ;
+}
```

Comment:
Why is `R` a trait parameter here? This is comparable to `Extend` or traits of that sort and it should instead be defined as such. It otherwise makes it hard to explicitely pass the type parameter if type hints are necessary.

```rust

pub
trait VecExtendFromReader {
    fn extend_from_reader<R : ReadIntoUninit> (
        self: &'_ mut Self,
        count: usize,
        reader: R,
    ) -> io::Result<()>
    ;
}
```

Reply:
Good catch!